internal/abci/abci.go

```diff
@@ -685,14 +680,73 @@ func (a *AbciApp) prepareMempoolTxns(txs [][]byte, maxBytes int, log *log.Logger
 
 		if bytes.Equal(tx.Sender, proposerAddr) {
 			proposerNonce = tx.Body.Nonce + 1
+			propTxs = append(propTxs, tx)
+		} else {
+			// Append the transaction to the final list.
+			// sender := string(tx.Sender)
+			otherTxns = append(otherTxns, tx)
+			// otherTxns[sender] = append(otherTxns[sender], tx)
 		}
-
-		finalTxns = append(finalTxns, txs[tx.is])
 		nonces = append(nonces, tx.Body.Nonce)
 		i++
 	}
 
-	return finalTxns, proposerNonce
+	// TODO: truncate based on our max block size since we'll have to set
+	// ConsensusParams.Block.MaxBytes to -1 so that we get ALL transactions even
+	// if it goes beyond max_tx_bytes.  See:
+	// https://github.com/cometbft/cometbft/pull/1003
+	// https://docs.cometbft.com/v0.38/spec/abci/abci++_methods#prepareproposal
+	// https://github.com/cometbft/cometbft/issues/980
+
+	// Enforce block size limits
+	// Txs order: MempoolProposerTxns, ProposerInjectedTxns, MempoolTxns
+	var finalTxs [][]byte
+
+	for _, tx := range propTxs {
+		txBts := txs[tx.is]
+		txSize := int64(len(txBts))
+		if maxTxBytes < txSize {
+			break
+		}
+		maxTxBytes -= txSize
+		finalTxs = append(finalTxs, txBts)
+	}
+
+	proposerTxs, err := a.txApp.ProposerTxs(context.Background(), proposerNonce, maxTxBytes, proposerAddr)
+	if err != nil {
+		log.Error("failed to get proposer transactions", zap.Error(err))
+	}
+
+	for _, tx := range proposerTxs {
+		txSize := int64(len(tx))
+		if maxTxBytes < txSize {
+			break
+		}
+		maxTxBytes -= txSize
+		finalTxs = append(finalTxs, tx)
+	}
+
+	// senders map is used to keep track of the senders that
+	// trasactions should be ignored from due to maxTxBytes limits
+	senders := make(map[string]bool)
+	for _, tx := range otherTxns {
+		sender := string(tx.Sender)
+		// If we have already added a transaction from this sender, skip it.
```

Comment:
To make sure I understand the intent, the comment means "If a previous transaction from this senders would have pushed over the bytes limit for the block, we skip all subsequent transactions from the sender because nonces need to be sequential. Keep checking transactions for other senders that may be smaller and fit in the remaining space."

Reply:
Yes, correct.